### PR TITLE
fix: remove unused time column when update dataset

### LINKED ISSUE #14683: [explore]Can't remove unnecessary Datetime column from time filter
Hello guys,
I have the issue that I can't remove unnecessary Datetime column from time filter while there are more than 1 Datetime column.  For example, there are two datetime columns named TradingTime and UpdateTime(which represent when to insert the record), only TradingTime is needed for time filter.

Steps to reproduce:

1. Add any datasource withmore than 1 Datetime column.
2. Explore the datasource, and note the time range is automatically applied as "Last week"
3. Edit datasource and uncheck "Is temporal/Is Filterable" with the column 'UpdateTime'
4. Viw quey sqls, 'UpdateTime' still in the where clause

Besides, I have tried another ways but  still not solved yet:

- Remove the column UpdateTime
- modify the type of column UpdateTime as String

### PR DESCRIPTION
**Pull request by zhaoyongjie**
### SUMMARY

closes: https://github.com/apache/superset/issues/14683
superset-ui related PR: https://github.com/apache-superset/superset-ui/pull/1148

Since the early Superset TableSource references to Druid table structure, there was always a time dimension(time partition column) in all tables(as known as, datasource or dataset). But this concept is not suitable for other data warehouses or databases. 

There is a `main_dttm_col` field in the superset dataset, service time dimension. Now if a time column is not passed to superset backend, the `main_dttm_col` will be used in the time column. 

This change 
- allows Superset to receive null time column
- force user update time column value when user changed existing time column type(for instance: uncheck is_temporal)

#### Before

https://user-images.githubusercontent.com/2016594/120643352-a3683f00-c4a8-11eb-9ed7-b9f3713236c9.mp4

#### After

https://user-images.githubusercontent.com/2016594/120643003-2b017e00-c4a8-11eb-9b11-59aac1384351.mp4

### ADDITIONAL INFORMATION

- [x] Has associated issue: https://github.com/apache/superset/issues/14683 https://github.com/apache/superset/issues/14754
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by junlincc** [comment]
Awesome PR description! thanks for the fix 🙏

File: superset-frontend/src/explore/components/controls/DatasourceControl/index.jsx [hunk 1]
```diff
@@ -117,6 +118,15 @@ class DatasourceControl extends React.PureComponent {
 
   onDatasourceSave(datasource) {
     this.props.actions.setDatasource(datasource);
+    // remove time column in the form_data
+    const timeCol = this.props.form_data?.granularity_sqla; // eslint-disable-line camelcase
```
Comment (villebro) [inline superset-frontend/src/explore/components/controls/DatasourceControl/index.jsx:122]: As there are multiple references to snake case variables, maybe we could just disable it for the whole file:
```js
/* eslint-disable-line camelcase */
```

File: superset-frontend/src/explore/components/controls/DatasourceControl/index.jsx [hunk 2]
```diff
@@ -117,6 +118,15 @@ class DatasourceControl extends React.PureComponent {
...
+    const { columns } = this.props.datasource;
+    if (
+      datasource.type === 'table' &&
+      !columns.find(({ column_name }) => column_name === timeCol)?.is_dttm // eslint-disable-line camelcase
+    ) {
+      this.props.actions.setControlValue('granularity_sqla', null);
+    }
```
Comment (villebro) [inline superset-frontend/src/explore/components/controls/DatasourceControl/index.jsx:129]: When the time column control is empty and a column is made `is_dttm`, the time column control stays empty, despite being required (not possible to remove value once set). Would it make sense to default to the first temporal column if the value is unset? Something like this:
```diff
diff --git a/superset-frontend/src/explore/components/controls/DatasourceControl/index.jsx b/superset-frontend/src/explore/components/controls/DatasourceControl/index.jsx
index 873067565..44452fb0c 100644
--- a/superset-frontend/src/explore/components/controls/DatasourceControl/index.jsx
+++ b/superset-frontend/src/explore/components/controls/DatasourceControl/index.jsx
@@ -121,12 +121,20 @@ class DatasourceControl extends React.PureComponent {
     // remove time column in the form_data
     const timeCol = this.props.form_data?.granularity_sqla; // eslint-disable-line camelcase
     const { columns } = this.props.datasource;
+    const firstDttmCol = columns.find(column => column.is_dttm);
     if (
       datasource.type === 'table' &&
+      timeCol &&
       !columns.find(({ column_name }) => column_name === timeCol)?.is_dttm // eslint-disable-line camelcase
     ) {
       this.props.actions.setControlValue('granularity_sqla', null);
+    } else if (datasource.type === 'table' && !timeCol && firstDttmCol) {
+      this.props.actions.setControlValue(
+        'granularity_sqla',
+        firstDttmCol.column_name,
+      );
     }
+
     if (this.props.onDatasourceSave) {
       this.props.onDatasourceSave(datasource);
     }
```

**Review by villebro — COMMENTED** [review]
A few minor ideas/comments

File: superset-frontend/src/explore/components/controls/DatasourceControl/index.jsx (hunk 2 above)
Comment (zhaoyongjie) [reply superset-frontend/src/explore/components/controls/DatasourceControl/index.jsx:129]: thanks for your comments!  This has the side effect of leaving an unexpected column to do the filtering and grain(the other words, first datetime column depends on the order of the columns in the database.). What do you think about?

Comment (zhaoyongjie) [reply superset-frontend/src/explore/components/controls/DatasourceControl/index.jsx]: Done this change. Thanks, comments!

**Review by villebro — APPROVED** [review]
LGTM

**Pull request merged**